Add DocumentDB support

jaywink opened the pull request:
Patch from https://github.com/agenda/agendash/pull/208 - merging in here in our fork given it is not yet merged by upstream. Have checked the code and tested it, seems to work so far.

slayer1551 committed:
Typo

slayer1551 committed:
Fix new job window modal from not closing

slayer1551 committed:
AWS Document DB Support

$facet is not available in AWS Document DB and throws an error.
Created an error check that if it fails with the standard query to try the AWS query to create the pages and filters.

slayer1551 committed:
Fix totalPage count

Fixed the total page count

slayer1551 committed:
Update agendash.js

removed line uses for testing

Half-Shot commented on public/app/js/confirms.js:
```diff
@@ -1,8 +1,8 @@
-   <div v-if="deletec" class="alert alert-success popupmessage">Job Deleted successfull</div>
-   <div v-else-if="requeuec" class="alert alert-success popupmessage">Job Requeue successfull</div>
-   <div v-else-if="createc" class="alert alert-success popupmessage">Job Created successfull</div>
+   <div v-if="deletec" class="alert alert-success popupmessage">Job Deleted successful</div>
```
Job Deleted successful*ly*, surely?

Half-Shot commented on lib/controllers/agendash.js:
```diff
@@ -56,69 +57,117 @@ module.exports = function (agenda, options) {
-            repeating: {
-              $and: ["$repeatInterval", { $ne: ["$repeatInterval", null] }],
-            },
+      },
+      {
+        $project: {
+          job: "$$ROOT",
+          _id: "$$ROOT._id",
+          running: {
+            $and: ["$lastRunAt", { $gt: ["$lastRunAt", "$lastFinishedAt"] }],
+          },
+          scheduled: {
+            $and: ["$nextRunAt", { $gte: ["$nextRunAt", new Date()] }],
+          },
+          queued: {
+            $and: [
+              "$nextRunAt",
+              { $gte: [new Date(), "$nextRunAt"] },
+              { $gte: ["$nextRunAt", "$lastFinishedAt"] },
+            ],
+          },
+          completed: {
+            $and: [
+              "$lastFinishedAt",
+              { $gt: ["$lastFinishedAt", "$failedAt"] },
+            ],
+          },
+          failed: {
+            $and: [
+              "$lastFinishedAt",
+              "$failedAt",
+              { $eq: ["$lastFinishedAt", "$failedAt"] },
+            ],
+          },
+          repeating: {
+            $and: ["$repeatInterval", { $ne: ["$repeatInterval", null] }],
```
The whitespace changes make this difficult to review

Half-Shot commented on lib/controllers/agendash.js:
```diff
@@ -56,69 +57,117 @@ module.exports = function (agenda, options) {
+        }
+      }
+    ]
+    const awsDocumentDBQuery = [
```
Can we explain the limitation with this query, for some reason we're trying to run a query, failing and then trying to run another query that's docDB compatible.

Half-Shot commented on lib/controllers/agendash.js:
```diff
@@ -56,69 +57,117 @@ module.exports = function (agenda, options) {
+    .then((result) => {
+      return result;
+    })
+    .catch((error) => { 
```
This is gross, it's an extra round trip to AWS every time we run this query, and I'm not sure amazon will thank us for that. Can we just make this a configuration option (or somehow smartly detect the version of Mongo, rather than this awful thing?)

We're not **even** looking at the error response to detect why it failed :face_exhaling:

jaywink replied:
I'm hoping to avoid us maintaining our own code until upstream implements this properly. This works and it's only doing this query every so and so seconds. I hardly think AWS will notice.

We can reimplement this patch but I'm going to postpone it after EMS staging is migrated.

jaywink replied:
I'm hoping to avoid us maintaining our own code until upstream implements this properly. This works and it's only doing this query every so and so seconds. I hardly think AWS will notice.

We can reimplement this patch but I'm going to postpone it after staging migration, if ok.

jaywink replied:
To clarify the reason for not spending engineering time now: when upstream implements documentdb support, we're going to want to use that implementation anyway, not whatever we came up with. As such I'd avoid making anything fancy at this stage. Given we only need documentdb support, I could simplify the pr - but that carries the issue if the upstream patch gets new commits, we'll be in a merge conflict against that.

Half-Shot reviewed (changes requested):
I feel like the fallback logic here is flawed.

jaywink committed:
Temp: Remove the attempt to do a $facet query

Make a proper patch where it is configurable later.

jaywink committed:
Temp: Apply a 1000 document limit to the jobs query

The DocumentDB query barfs on large data sets. This is a temporary fix until the documentdb query is properly fixed.
